l10n_it_delivery_note/models/stock_delivery_note.py

```diff
@@ -900,6 +900,71 @@ def get_location_address(self, location_id):
 
         return location_address
 
+    def _grouped_lines(self):
+        def create_groups(show_so):
+            lines = self.env["stock.delivery.note.line"]
+            group_prod = self.line_ids.mapped("product_id")
+            for prod in group_prod:
+                filtered_lines = self.line_ids.filtered(lambda x: x.product_id == prod)
+                # for the case in which there are more discuonts, price_unit, taxes
+                if len(set(filtered_lines.mapped("discount"))) > 1:
+                    lines += filtered_lines
+                    continue
+                if len(set(filtered_lines.mapped("price_unit"))) > 1:
+                    lines += filtered_lines
+                    continue
+                tax_ids = filtered_lines.mapped("tax_ids")
+                different_taxes = False
+                for line in filtered_lines:
+                    if not line.tax_ids == tax_ids:
+                        lines += filtered_lines
+                        different_taxes = True
+                        break
+                if different_taxes:
+                    continue
+                if show_so:
+                    so_numbers = {line.sale_order_number for line in filtered_lines}
```

Comment:
**issue**: if `show_so` is False the variable `so_numbers` is never declared, so its use at line 950 would raise an exception.

**suggestion**: we could replace these two lines with one-liner that also solves the issue above
```suggestion
                so_numbers = {line.sale_order_number for line in filtered_lines} if show_so else {}
```